### ISSUE TITLE
A dataset of large Mega Man level scenes

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#148]
One of the big challenges in training on Mega Man level scenes is how the levels zig-zag around, moving up, down, left, right as opposed to just left to right. Part of the challenge of this could possibly be ignored if we just trained on views of large portions of the levels, kind of like what we are doing with Mario in issue #96 . Look at the Mega Man data from TheVGLC ( https://github.com/schrum2/TheVGLC ) and see if we can extract a dataset of larger level scenes. Maybe 64 by 64? If any levels are narrower/shorter than that, we can pad the scene with the @ character, which represents out of bounds level portions. There already exists code in the repo for processing Mega Man level data (it understands the tile set), so adapting this to get larger scenes should be possible (though probably not as straight forward as with Mario).

Create this dataset, and then check it out in the ascii data browser to see what it is like. I'm not too worried about captions at the moment, since the first thing we would want to try with this data is training an unconditional diffusion model. Depending on how the unconditional model training goes, we can decide what to do next.

This is a step to scaling up to use Mega Man Maker data ( #147 )

**Comment by fwilson12** [comment]
Was able to generate 32x32 scenes simply with CLI args to the existing script; generating 64x64 samples was inconsistent and led to index OOB exceptions, and nearly every scene was mostly out of bounds. 

### 32^2 scenes

<img width="599" height="589" alt="Image" src="https://github.com/user-attachments/assets/712bcda1-2d08-4767-860b-d9570459f9c5" />

<img width="539" height="547" alt="Image" src="https://github.com/user-attachments/assets/a9446b64-099e-4605-bd51-4d9fe71d0b62" />

<img width="566" height="559" alt="Image" src="https://github.com/user-attachments/assets/fe00e2ea-3018-420a-b790-d044765bfea8" />

### 64^2 scenes

<img width="679" height="659" alt="Image" src="https://github.com/user-attachments/assets/98ab8c80-0cc9-42ca-97e2-7b3f9d4122d3" />

<img width="655" height="651" alt="Image" src="https://github.com/user-attachments/assets/df63acc5-7489-4099-81b6-bd5862f918f5" />

<img width="656" height="658" alt="Image" src="https://github.com/user-attachments/assets/acbb33a0-ee50-4d00-b526-7d632b0767d1" />

**Generating 64x64 scenes also produced some errant scenes of different size:**

<img width="665" height="231" alt="Image" src="https://github.com/user-attachments/assets/2ce6c13d-9455-4676-98d1-02c83a9a9d97" />

**Comment by schrum2** [comment]
Some clarification to @fwilson12 about the Mega Man data. Most instructions for how to work with Mega Man are in 
https://github.com/schrum2/MarioDiffusion/blob/main/MM_README.md
Also look in:
https://github.com/schrum2/MarioDiffusion/tree/main/MM_Batch
Notice that if you run 
https://github.com/schrum2/MarioDiffusion/blob/main/MM_Batch/MM-data.bat
you get several files. In particular, it seems that the generated files MM_Levels_Full.json and MM_Levels_Simple.json use the json property value "sample" whereas the generated files "MM_LevelsAndCaptions-full-regular.json" and "MM_LevelsAndCaptions-simple-regulsr.json" have the json property value "scene". So, it should be possible to work with the right files without manually changing property values from "sample" to "scene"

**Comment by schrum2** [comment]
We have a collection of larger scenes. Let's try training on it. Go ahead and make a batch file for testing this out so that you can share it with me, and I'll run it over the weekend.

**Comment by fwilson12** [comment]
As it stood before, creating larger MM datasets with create_megaman_json_data.py was basically impossible with the way the collision/null char detection worked when snaking across vertical and horizontal corridors. This meant each MM level would only yield 2 or 3 scenes each. After finagling and help from claude, create_megaman_json_data.py in my branch now still navigates levels based on the original 16x14 scene navigation/edge detection, but can capture larger level scenes. I'm not sure if this is the best approach, but regardless I began training an unconditional model using the larger 32x32 dataset (see commands below).

- python create_megaman_json_data.py  --output datasets\MM_Levels-simple32.json --target_height 32 --target_width 32 --group_encodings

- python MM_create_ascii_captions.py --dataset datasets\MM_Levels-simple32.json --tileset datasets\MM_Simple_Tileset.json --output datasets\MM_LevelsAndCaptions-simple32-regular.json

- python tokenizer.py save --json_file datasets\MM_LevelsAndCaptions-simple32-regular.json --pkl_file datasets\MM_Tokenizer-simple32-regular.pkl

- python create_random_test_captions.py --save_file datasets\MM_RandomTest_simple32-regular.json --json datasets\MM_LevelsAndCaptions-simple32-regular.json --seed 0 --game MM-Simple

- python split_data.py --json_file datasets\MM_LevelsAndCaptions-simple32-regular.json --train_pct .9 --val_pct .05 --test_pct .05  --seed 0 --game mm-simple

- python train_diffusion.py --game MM-Simple --augment --output_dir MM-simple32-unconditional0 --num_epochs 500 --json datasets\MM_LevelsAndCaptions-simple32-regular-train.json --val_json datasets\MM_LevelsAndCaptions-simple32-regular-validate.json --seed 0

**Comment by fwilson12** [comment]
I committed the changes above to my branch; the original ceiling padding approach when target height/width are not specified still works exactly as before, so the changes made should only be applied when sampling larger scenes.

**Comment by fwilson12** [comment]
Training is done, and apart from some connectedness issues, the samples don't look too bad:

##

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/44aa3d59-3257-4403-aa34-1b781adb7a1e" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/7e02143d-0a9d-441d-9072-2c3c3f42bbfb" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/d014def4-9846-4d52-9e91-a47f073c0a3e" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/23bd238b-faeb-435c-be6a-10f8b6cacd6c" />

##

An interesting feature of the generated samples is that they seemed to generalize out-of-bounds regions; the model was also able to produce levels with playable regions greater than those found in its training data, and some generated levels fill out nearly the entire 32x32 grid. There's still a ton of variables regarding data collection and sampling, and some of the scenes it's generating couldn't appear in an actual level.

I think an Interesting next step is to train on a dataset with even larger scenes, and see if it could produce coherent level chunks, and correctly navigate the balance between the horizontal/vertical corridors and the surrounding out-of-bounds regions.

**Comment by fwilson12** [comment]
Just made a dataset of 64x64 level chunks and started training a model on it; might end up having to reduce the batch size to 16 if the batch tensors end up being too large

**Comment by schrum2** [comment]
I went ahead and quickly trained my own unconditional model using the instructions in the post above that starts with "As it stood before" and got these results:

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/e4d06603-ed33-4be5-8c94-04b43f868eed" />
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/0ead41c6-0ab3-480b-9d38-129a3a02eac7" />
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/8657d7fb-e234-4760-96f0-b04e2308d47a" />
<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/3ed0ba49-2b0e-4a53-8246-67564e3ad69b" />

I think that the results look good, but sometimes fail on functional requirements such as traversability, but I think that might be the fault of the training data. I believe that scanning it the way you are probably creates scenes with inaccessible portions.

This gives me some new ideas related to how we create training sets.

After you post the results from your run that is going over the weekend, I think you can close this issue, but I'm going to make another issue for you inspired by this.

**Comment by fwilson12** [comment]
The results from the 64x64 model are decent, but again often feature impossible-to-traverse segments which I think can be safely attributed to the data it was trained on, which contains come 64x64 snapshots with parts of the traversable corridors truncated:

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/b4d453ca-f4bf-4d05-9e6d-97d8364a5161" />

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/088cf3eb-9daf-4c5a-98ae-7dec5f858b05" />

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/3274c533-64a4-43cf-bb50-8705007763cb" />

 ##

However, some of the samples do look quite good, so if we can somehow filter out untraversable scenes from the training data or incorporate a traversability measure into the loss function during training, perhaps we can achieve even better results:

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/4c87224a-004a-412e-994d-b8efbb31c740" />

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/62c0174a-c731-4c6f-b7a0-f7cc422ca7d6" />

##

I'll go ahead and close the issue, but work related to this thread will be expanded upon in newer issues (#164)

**Comment by schrum2** [comment]
I really like the results you posted above, and I have a new idea, so I'm re-opening the issue. Some of those scenes definitely look good. I'm not as familiar with the Mega Man levels as I am with Mario, so can you please check to see if some of the really good model outputs are basically just direct copies of training data?

More importantly, I think the main source of traversability problems is the fact that the sliding window for creating the training data sometimes catches thin edges of a room that exclude the part of it that actually connects to the adjacent area.

Try creating a training set where the stride of the sliding window used to create the training set matches the dimensions of what is visible on the screen at any given time. For Mega Man, I'm pretty sure this means we have a horizontal stride of 16 and a vertical stride of 14 (double check that). You may need to add code to support different stride amounts along each dimension. The idea is to make sure that the edges of the training samples always align with the edges of rooms.

You'll be working on #164 as well, but see if you can get this training soon, in case simply retricting the data the right way solves the problem. Given the reduced stride, this should be a much smaller training set ... could have negative consequences.

**Comment by fwilson12** [comment]
After looking through the training data again, the scenes above are pretty much carbon copies of select scenes from the training data, similar to what we've seen with training models on larger Mario levels. The adjusted stride approach should rectify the connectedness issues though, I'll tackle this after I fix the 3x3 tile creation related to the block embedding issue.

**Comment by schrum2** [comment]
I assume we are universally training for about 500 epochs on these, just because that was the norm with Mario data. For this large dataset, try reducing that to 300 epochs. It will end sooner but also might overfit less

**Comment by fwilson12** [comment]
Due to the current way the create_mm_level script scans levels from the vglc, it's tricky to avoid capturing level slices that are un-traversable. Thus, an intuitive next step is to factor in A* traversability as a filter to discard these un-traversable scenes. This should also help with the dataset size issue, as we won't have to scan through the level in larger discrete chunks. Will work on this in issue #164 and come back when progress has been made there.

**Comment by schrum2** [comment]
When creating a Mega Man dataset, add an option to filter out any levels that don't have a solvable A* path. From what we've seen so far, this will exclude a few levels that actually are beatable due to bad exit orb placement, but that hopefully won't matter much. Track stats on how many potential level scenes are discarded as a results of this filtering. Let's see if we can make a good dataset of large levels, and then train on it.

**Comment by fwilson12** [comment]
Tagged the commit for this issue under #164, but this is now in place; I have an unconditional model training on a filtered 32x32 dataset  (contains ~70% of the original samples, which is around 1800). Hopefully the results will be noticeably better

**Comment by fwilson12** [comment]
Model above is done training, and the results aren't what I hoped; not only do the samples themselves look pretty rough, 59% of the batch of 100 samples I had it generate were considered un-traversable by my spawn/orb placement A* heuristic despite the fact that each sample in the dataset it trained on WAS considered traversable by the same algorithm. 



## Examples (green dot: spawn, blue dot: orb)

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/588593e1-5088-42c2-b2ff-f9f4edb45a56" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/d44bed10-92da-46c9-a11e-f37f367f4e8a" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/22329f20-24f8-4975-922d-a2e0142b01d6" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/15891f27-6d88-4023-be20-de01fb74002e" />

##

It's worth noting that the orb/spawn placement heuristic is at fault for a few of the samples, but for the most part the generated level structure was the cause of the unsolvability; will keep this issue open for now if we decide to look for a better way to sample larger scenes from MM levels, but I think the more natural step would be to somehow incorporate the traversability measure into training in some way.

**Comment by schrum2** [comment]
Try again with a conditional model

**Comment by fwilson12** [comment]
Conditional model is training, but it's not predicted to finish training until later this week; we haven't trained a conditional model on larger MM datasets yet, but it seems the training time increase is comparable to what we experienced with the larger mario models. 

I'll make a batch file as well

**Comment by fwilson12** [comment]
.bat in the commit above is located in the MM_batch dir

**Comment by schrum2** [comment]
I'm running `MM_conditional-filtered.bat 32` at home

**Comment by schrum2** [comment]
The results of this actually look reasonably strong to me, though obviously not fully accurate:

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/ce4769df-f913-4468-aa8f-e4efae8fa92c" />

full floor. two enemies. exit direction right. entrance direction left. all water. two rectangular block clusters.

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/24651b1e-24de-465e-b92e-762332d76e05" />

one ladder. entrance direction down. giant gap with two chunks of floor. exit direction up. two irregular block clusters.

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/78623e26-aa0c-4f13-8495-c2226d2eb525" />

two loose blocks. one enemy. two platforms. one irregular block cluster. floor with one gap. exit direction right. two rectangular block clusters. entrance direction left.

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/7cd02c97-2503-4862-bccf-e98ee4eab9aa" />

a few platforms. one rectangular block cluster. entrance direction up. giant gap with one chunk of floor. several hazards. one enemy. a few powerups. one loose block. two ladders. exit direction down. a few irregular block clusters.

--------------------------------------------------------------------------------------------

Of course, it's a bit luck of the draw regarding which captions are sampled.

I loaded the model into the interactive generator and results were much worse.

<img width="599" height="679" alt="Image" src="https://github.com/user-attachments/assets/edc66a9a-defe-40bd-be36-5e6a63da129f" />

<img width="565" height="693" alt="Image" src="https://github.com/user-attachments/assets/d8dfc320-73f2-47a7-a30f-c299ae0b94b5" />

Despite that, I think this issue can be closed
